File: src/roleplay_catalogue/misc/password_strength.py
import re


PASSWORD_MIN_LENGTH = 8
PASSWORD_MAX_LENGTH = 128

_LOWERCASE = re.compile(r'[a-z]')
_UPPERCASE = re.compile(r'[A-Z]')
_DIGIT = re.compile(r'[0-9]')
_SPECIAL = re.compile(r'[^A-Za-z0-9]')


def password_strength_error(password: str) -> str | None:
    """Return a description of the first unmet password complexity rule, or None if it passes.

    Length is intentionally not checked here; callers already enforce
    PASSWORD_MIN_LENGTH/PASSWORD_MAX_LENGTH via Pydantic Field constraints.
    """
    if not _LOWERCASE.search(password):
        return 'Password must contain at least one lowercase letter'
    if not _UPPERCASE.search(password):
        return 'Password must contain at least one uppercase letter'
    if not _DIGIT.search(password):
        return 'Password must contain at least one number'
    if not _SPECIAL.search(password):
        return 'Password must contain at least one special character'
    return None

File: frontend/src/utils/passwordStrength.js
export const PASSWORD_MIN_LENGTH = 8
export const PASSWORD_MAX_LENGTH = 128

// Mirrors the backend policy in src/roleplay_catalogue/misc/password_strength.py: 8-128
// characters, at least one lowercase letter, one uppercase letter, one number, and one
// character outside A-Za-z0-9. Keep both in sync if the policy ever changes.
const RULES = [
  { key: 'tooShort', test: (password) => password.length >= PASSWORD_MIN_LENGTH },
  { key: 'tooLong', test: (password) => password.length <= PASSWORD_MAX_LENGTH },
  { key: 'missingLowercase', test: (password) => /[a-z]/.test(password) },
  { key: 'missingUppercase', test: (password) => /[A-Z]/.test(password) },
  { key: 'missingDigit', test: (password) => /[0-9]/.test(password) },
  { key: 'missingSpecial', test: (password) => /[^A-Za-z0-9]/.test(password) },
]

// Returns the translation key for the first unmet rule, or null if the password satisfies
// the whole policy. Checked in the same order as the backend so the reported reason matches.
export function passwordStrengthError(password) {
  const failed = RULES.find((rule) => !rule.test(password))
  return failed ? failed.key : null
}
